File: cmd/add.go
package cmd

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"

	"github.com/serhiitroinin/hostler/internal/config"
	"github.com/serhiitroinin/hostler/internal/hosts"
	"github.com/serhiitroinin/hostler/internal/nginx"

	"github.com/fatih/color"
	"github.com/spf13/cobra"
)

// getExecutablePath returns the absolute path to the current executable
func getExecutablePath() string {
	execPath, err := os.Executable()
	if err != nil {
		return "hostler"
	}
	// Resolve symlinks to get the real path
	if realPath, err := filepath.EvalSymlinks(execPath); err == nil {
		return realPath
	}
	return execPath
}

var addCmd = &cobra.Command{
	Use:   "add <domain> <port>",
	Short: "Add a new local domain",
	Long: `Add a new local domain pointing to a specific port.

Example:
  hostler add myapp.loc 3000
  hostler add api.loc 8080`,
	Args: cobra.ExactArgs(2),
	Run:  runAdd,
}

func runAdd(cmd *cobra.Command, args []string) {
	domain := args[0]
	port, err := strconv.Atoi(args[1])
	if err != nil || port < 1 || port > 65535 {
		color.Red("Error: Invalid port number: %s", args[1])
		os.Exit(1)
	}

	// Check if hostler has been initialized
	if !config.IsInitialized() {
		color.Red("Error: hostler not initialized.")
		fmt.Println()
		fmt.Println("Run 'sudo hostler init' to set up hostler for passwordless operation.")
		os.Exit(1)
	}

	userConfigDir := config.GetCurrentUserConfigDir()

	fmt.Println()
	color.Cyan("Detecting nginx configuration...")

	// Detect nginx
	cfg, err := nginx.Detect()
	if err != nil {
		color.Red("Error: %v", err)
		os.Exit(1)
	}

	fmt.Printf("  nginx version: %s\n", cfg.Version)
	fmt.Printf("  config dir: %s\n", userConfigDir)

	// Check if nginx is running
	if !cfg.IsRunning {
		color.Yellow("Warning: nginx is not running")
		fmt.Print("  Start nginx? [y/N]: ")
		var answer string
		fmt.Scanln(&answer)
		if answer == "y" || answer == "Y" {
			if err := nginx.Start(); err != nil {
				color.Red("Error: %v", err)
				os.Exit(1)
			}
			color.Green("  nginx started")
		}
	} else {
		color.Green("  nginx is running")
	}

	fmt.Println()
	color.Cyan("Checking for conflicts...")

	// Read existing entries from user config directory
	entries, err := nginx.ParseUserConfigs(userConfigDir)
	if err != nil {
		color.Red("Error: Failed to parse configs: %v", err)
		os.Exit(1)
	}

	// Check for conflicts in our managed configs
	for _, entry := range entries {
		if entry.Domain == domain {
			color.Red("Error: Domain '%s' already exists (port %d)", domain, entry.Port)
			color.Yellow("  Use 'hostler remove %s' first, or choose a different domain", domain)
			os.Exit(1)
		}
		if entry.Port == port {
			color.Red("Error: Port %d is already used by '%s'", port, entry.Domain)
			color.Yellow("  Use a different port or remove the existing domain first")
			os.Exit(1)
		}
	}

	// Check for conflicts in system nginx configs
	domainConflict, portConflict, err := nginx.FindConflicts(cfg.IncludeDir, domain, port, true)
	if err != nil {
		color.Yellow("Warning: Could not check for conflicts: %v", err)
	}
	if domainConflict != "" {
		color.Red("Error: Domain '%s' already configured in: %s", domain, domainConflict)
		os.Exit(1)
	}
	if portConflict != "" {
		color.Yellow("Warning: Port %d may conflict with config in: %s", port, portConflict)
	}

	// Check hosts file
	exists, err := hosts.HasDomain(hosts.GetHostsPath(), domain)
	if err != nil {
		color.Yellow("Warning: Could not check hosts file: %v", err)
	}
	if exists {
		color.Yellow("Warning: Domain '%s' already exists in /etc/hosts (will be updated)", domain)
	}

	color.Green("  No conflicts found")

	fmt.Println()
	color.Cyan("Updating configuration...")

	// Write nginx config to user directory (no sudo needed)
	if err := nginx.WriteUserDomainConfig(userConfigDir, domain, port); err != nil {
		color.Red("Error: Failed to write nginx config: %v", err)
		os.Exit(1)
	}
	color.Green("  Created %s/%s.conf", userConfigDir, domain)

	// Add to hosts file via sudo (passwordless via sudoers)
	hostlerPath := getExecutablePath()
	hostsCmd := exec.Command("sudo", hostlerPath, "_hosts-add", domain)
	if output, err := hostsCmd.CombinedOutput(); err != nil {
		color.Red("Error: Failed to update hosts file: %v", err)
		if len(output) > 0 {
			fmt.Println(string(output))
		}
		// Rollback: remove the config file we just created
		nginx.RemoveUserDomainConfig(userConfigDir, domain)
		os.Exit(1)
	}
	color.Green("  Updated /etc/hosts (via sudo)")

	// Test nginx config via sudo
	testCmd := exec.Command("sudo", "nginx", "-t")
	if output, err := testCmd.CombinedOutput(); err != nil {
		color.Red("Error: nginx config test failed")
		fmt.Println(string(output))
		color.Yellow("  Rolling back changes...")
		// Rollback: remove config and hosts entry
		nginx.RemoveUserDomainConfig(userConfigDir, domain)
		exec.Command("sudo", hostlerPath, "_hosts-remove", domain).Run()
		os.Exit(1)
	}
	color.Green("  nginx config is valid")

	// Reload nginx via sudo
	if cfg.IsRunning {
		reloadCmd := exec.Command("sudo", "nginx", "-s", "reload")
		if output, err := reloadCmd.CombinedOutput(); err != nil {
			color.Red("Error: Failed to reload nginx")
			fmt.Println(string(output))
			os.Exit(1)
		}
		color.Green("  nginx reloaded")
	}

	fmt.Println()
	color.Green("Successfully added %s -> localhost:%d", domain, port)
	fmt.Printf("\n  Access your app at: http://%s\n\n", domain)
}

File: src/commands/add.ts
// `hostler add <domain> <port>` — create or update a domain → port mapping.
//
// Re-adding an existing domain with a new port performs an atomic UPDATE rather
// than erroring. nginx config writes go through the privileged `_nginx-add`
// helper (the config dir is root-owned), and every failure path — hosts update,
// `nginx -t`, and reload — rolls the change back to the prior state.
import { isInitialized, getCurrentUserConfigDir } from "../lib/config.ts";
import { isValidDomain, normalizeDomain, parsePort } from "../lib/domain.ts";
import { run, selfInvocation } from "../lib/exec.ts";
import * as hosts from "../lib/hosts.ts";
import * as nginx from "../lib/nginx.ts";
import { confirm, green, printError, printInfo, printOk, printWarn } from "../lib/ui.ts";

export async function add(args: string[]): Promise<void> {
  if (!args[0] || !args[1]) {
    printError("Usage: hostler add <domain> <port>");
    process.exit(1);
  }

  const domain = normalizeDomain(args[0]);
  const port = parsePort(args[1]);

  if (port === null) {
    printError(`Invalid port number: ${args[1]}`);
    printWarn("  Port must be an integer between 1 and 65535");
    process.exit(1);
  }
  if (!isValidDomain(domain)) {
    printError(`Invalid domain format: ${args[0]}`);
    printWarn("  Domains must look like 'myapp.loc' (letters, numbers, hyphens, at least one dot)");
    process.exit(1);
  }

  if (!isInitialized()) {
    printError("hostler not initialized.");
    console.log("\nRun 'sudo hostler init' to set up hostler for passwordless operation.");
    process.exit(1);
  }

  const configDir = getCurrentUserConfigDir();
  const sudoSelf = (...a: string[]) => run(["sudo", ...selfInvocation(), ...a]);

  console.log();
  printInfo("Detecting nginx configuration...");

  let cfg: nginx.NginxConfig;
  try {
    cfg = await nginx.detect();
  } catch (err) {
    printError(message(err));
    process.exit(1);
  }

  console.log(`  nginx version: ${cfg.version}`);
  console.log(`  config dir: ${configDir}`);

  if (!cfg.isRunning) {
    printWarn("Warning: nginx is not running");
    if (await confirm("  Start nginx? [y/N]: ")) {
      try {
        await nginx.start();
        printOk("  nginx started");
        cfg.isRunning = true;
      } catch (err) {
        printError(message(err));
        process.exit(1);
      }
    }
  } else {
    printOk("  nginx is running");
  }

  console.log();
  printInfo("Checking for conflicts...");

  const entries = await nginx.parseUserConfigs(configDir);
  const existing = nginx.findEntry(entries, domain);

  if (existing && existing.port === port) {
    printOk(`  ${domain} is already mapped to localhost:${port} — nothing to do`);
    console.log();
    return;
  }

  const isUpdate = existing !== undefined;

  const portTwin = entries.find((e) => e.domain !== domain && e.port === port);
  if (portTwin) {
    printWarn(`Warning: port ${port} is also used by '${portTwin.domain}'`);
  }

  // Another user's hostler configs live under a sibling /etc/hostler/<user>/ dir;
  // nginx would only warn ("conflicting server name") and silently ignore the
  // duplicate, so catch it here as a hard error.
  if (!isUpdate) {
    const otherUserDir = await nginx
      .findDomainInOtherUserDirs(domain, configDir)
      .catch(() => null);
    if (otherUserDir) {
      printError(`Domain '${domain}' is already configured by another user in: ${otherUserDir}`);
      process.exit(1);
    }
  }

  try {
    const { domainConflict, portConflict } = await nginx.findConflicts(cfg.includeDir, domain, port);
    if (domainConflict) {
      printError(`Domain '${domain}' is already configured in: ${domainConflict}`);
      process.exit(1);
    }
    if (portConflict) {
      printWarn(`Warning: port ${port} may conflict with config in: ${portConflict}`);
    }
  } catch (err) {
    printWarn(`Warning: Could not check for conflicts: ${message(err)}`);
  }

  if (!isUpdate) {
    const inHosts = await hosts.hasDomain(hosts.getHostsPath(), domain).catch(() => false);
    if (inHosts) {
      printWarn(`Warning: Domain '${domain}' already exists in /etc/hosts (will be updated)`);
    }
  }

  printOk("  No conflicts found");

  console.log();
  printInfo("Updating configuration...");

  // Restores the prior state. On update, rewrite the previous port (when known);
  // otherwise tear down the new config and hosts entry.
  const rollback = async () => {
    if (isUpdate && existing!.port !== null) {
      await sudoSelf("_nginx-add", domain, String(existing!.port));
    } else {
      await sudoSelf("_nginx-remove", domain);
      if (!isUpdate) await sudoSelf("_hosts-remove", domain);
    }
  };

  // 1. Write the nginx config as root via the privileged helper.
  const writeRes = await sudoSelf("_nginx-add", domain, String(port));
  if (!writeRes.ok) {
    printError("Failed to write nginx config");
    if (writeRes.combined.trim()) console.log(writeRes.combined.trim());
    process.exit(1);
  }
  if (isUpdate) printOk(`  Updated ${configDir}/${domain}.conf (port ${existing!.port ?? "?"} → ${port})`);
  else printOk(`  Created ${configDir}/${domain}.conf`);

  // 2. Add the hosts entry (idempotent; already present on update).
  const hostsRes = await sudoSelf("_hosts-add", domain);
  if (!hostsRes.ok) {
    printError("Failed to update hosts file");
    if (hostsRes.combined.trim()) console.log(hostsRes.combined.trim());
    await rollback();
    process.exit(1);
  }
  if (!isUpdate) printOk("  Updated /etc/hosts (via sudo)");

  // 3. Validate. `nginx -t` exits 0 even when it merely *warns* about a
  // conflicting server name (a duplicate from any source — another user, a stale
  // config). Treat such a warning for our domain as a hard failure, since nginx
  // would otherwise silently ignore one of the duplicates.
  const nginxBin = await nginx.resolveNginxBin();
  const test = await run(["sudo", nginxBin, "-t"]);
  const conflictsOurDomain = new RegExp(
    `conflicting server name "${escapeRegExp(domain)}"`,
  ).test(test.combined);
  if (!test.ok || conflictsOurDomain) {
    printError(conflictsOurDomain ? `Domain '${domain}' conflicts with an existing server block` : "nginx config test failed");
    console.log(test.combined.trim());
    printWarn("  Rolling back changes...");
    await rollback();
    process.exit(1);
  }
  printOk("  nginx config is valid");

  // 4. Reload — and roll back the running config if the reload fails.
  if (cfg.isRunning) {
    const reload = await run(["sudo", nginxBin, "-s", "reload"]);
    if (!reload.ok) {
      printError("Failed to reload nginx");
      console.log(reload.combined.trim());
      printWarn("  Rolling back changes...");
      await rollback();
      await run(["sudo", nginxBin, "-s", "reload"]); // best-effort revert of running state
      process.exit(1);
    }
    printOk("  nginx reloaded");
  }

  console.log();
  const verb = isUpdate ? "Updated" : "Successfully added";
  printOk(`${verb} ${domain} -> localhost:${port}`);
  console.log(`\n  Access your app at: ${green(`http://${domain}`)}\n`);
}

function message(err: unknown): string {
  return err instanceof Error ? err.message : String(err);
}

function escapeRegExp(s: string): string {
  return s.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}
